---
title: Comparison of Filtering Cut-Offs
author: CRISPR Functional Genomics
output:
  bookdown::html_document2:
    theme: journal
    highlight: kate
    toc: yes
    toc_depth: 3
    toc_float:
      collapsed: no
date: "`r format(Sys.time(), '%d %B, %Y')`"
---



Run by: `r Sys.getenv("USER")` 


 The rationale and process:

* Not all theoretical combinations of sgRNA - RSL barcodes are present in the input library; Our goal is to detect the combinations which are artefacts arisen due to sequencing errors; These artefacts, which typically are present in only one technical replicate and / or are detected by very small number of read counts give rise to the zero inflated signal in downstream RSL analysis;

* Input library has been sequenced in two technical replicates to allow for detection of these artefactual reads;

* sgRNA - RSL combinations are retained if they are present in **both replicates** and have a **read count >= cutoff** in both replicates individually;

* Filtering cutoffs in the range of 0 (no filtering) to 10 (excessive filtering) are compared with respect to the following metrics:

  * number of retained sgRNA - RSL combinations in the filtered input library;

  * library sizes of the input library technical replicates after the filtering;


* Finally, each of the input libraries filtered using different cutoffs is used for processing a reference data set and the following metrics are evaluated:

  * number of detected sgRNA - RSL combinations in the reference data set after filtering;

  * number of detected sgRNAs in the reference data set after filtering;;

  * library sizes of the reference data set samples after filtering;

  * distribution of read counts per sgRNA - RSL combination (histogram plots);

  * distribution of RSL counts per sgRNA (histogram and box plots);

  * comparison of RSL counts per sgRNA between different strategies (scatter plots);



----------------

```{r, echo=FALSE}
```


```{r local-tsts, include=FALSE,cache=FALSE}
#path to count table for reference data
#refdata.path="/Users/agata.smialowska/NBISproj/5351_CRISPR_pipeline/CRISPR_proj_cont/tests/input/data/Calle_Heldin.UMIcounts.csv"

#root dir for filtered input libraries
#inputdata.path="/Users/agata.smialowska/NBISproj/5351_CRISPR_pipeline/CRISPR_proj_cont/tests/input/data/Brunello_v2022/results/input_filtered"

#inputlib.name="Brunello_v2022"

#refdata.pref="Calle_Heldin.UMIcounts"

#refdata.defined=TRUE
#refdata.defined=FALSE



```


```{r include=FALSE,cache=FALSE}
library(knitr)
library(bookdown)


knitr::opts_chunk$set(echo = FALSE, 
                      collapse = TRUE, 
                      warning = FALSE, 
                      message = FALSE)

```




```{r dirs, include=FALSE,cache=FALSE}


```


```{r prep_environment, include=FALSE,cache=FALSE}
library(ggplot2)
library(tibble)
library(tidyr)
library(dplyr)
library(reshape2)
library(viridis)
library(parallel)
#library(cowplot)
library(gridExtra)

library(kableExtra)


refdata.status=isTRUE(refdata.defined=="TRUE")
```

# Analysis of Input Library `r inputlib.name`


```{r input-stats, include=FALSE,cache=FALSE}
fname_log=paste0(inputlib.name,".report.log")

resdirs=dir(inputdata.path)

#exclude dirs with refdata.pref
resdirs=resdirs[!grepl(refdata.pref, resdirs)]

#input.stats <- vector(mode = "list", length = length(resdirs))

input.stats <- vector(mode = "list")


for (i in resdirs){
  input.stats.i=file.path(inputdata.path,i,fname_log)
  print("Processing")
  print(input.stats.i)

  input.metrics.i=read.table(input.stats.i, sep="\t", header=FALSE, blank.lines.skip=TRUE)

  input.stats[[i]]=input.metrics.i

}


input.stats.df <- data.frame(matrix(unlist(input.stats), nrow=length(input.stats), byrow=TRUE))
colnames(input.stats.df)=c("CO","sgRNA.RSL","ColSums")
input.stats.df$CO=as.numeric(as.character(input.stats.df$CO))
input.stats.df=input.stats.df[order(input.stats.df$CO,decreasing=FALSE),]

input.stats.df$sgRNA.RSL=format(as.numeric(input.stats.df$sgRNA.RSL), nsmall=0, big.mark=",")
colnames(input.stats.df)=c("CO","sgRNA-RSL","ColSums")

input.stats.df$Rep1=gsub("\\,.*","",input.stats.df$ColSums)
input.stats.df$Rep2=gsub(".*,","",input.stats.df$ColSums)

input.stats.df$Rep1=format(as.numeric(input.stats.df$Rep1), nsmall=0, big.mark=",")
input.stats.df$Rep2=format(as.numeric(input.stats.df$Rep2), nsmall=0, big.mark=",")

input.stats.df=input.stats.df[,-3]


```

```{r input-stats-table, include=TRUE,cache=FALSE, results='asis'}

title.1="Number of detected sgRNA-RSL combination under different filtering cut-off settings."

footnote.1="Number of sgRNA-RSL combinations and library sizes are calculated after the filtering of the input library using the indicated cutoff."

kable(input.stats.df, align = c("l", "c", "c", "c"), booktabs = TRUE, row.names = FALSE, caption = title.1, 
  col.names=c("Cut Off","Number of sgRNA-RSL Combinations","Library Size (technical replicate 1)","Library Size (technical replicate 2)") )%>%
    footnote(
      general = footnote.1,
      general_title = "Note.",
      footnote_as_chunk = TRUE
      )%>%
  kable_minimal(full_width = TRUE)

```



```{r input-stats-hist, include=FALSE,cache=FALSE}

fname_freq=paste0(inputlib.name,".frequencies.raw_reads_aggregated.tsv")

freq.stats <- vector(mode = "list", length=length(resdirs))

freq.hist <- vector(mode = "list")


for (i in resdirs){
  freq.stats.i=file.path(inputdata.path,i,fname_freq)
  print("Processing")
  print(freq.stats.i)

  freq.metrics.i=read.table(freq.stats.i, sep="\t", header=FALSE, blank.lines.skip=TRUE)
  #colnames(freq.metrics.i)=c("sgRNA.RSL.readcount","frequency")
  colnames(freq.metrics.i)=c("readcount","frequency")
  freq.metrics.i$CO=i

  freq.stats[[i]]=freq.metrics.i

}

freqs=bind_rows(freq.stats)

title_hist_fs="Histogram of sgRNA-RSL readcount frequncies in Input library under different filiering cut-offs. Full Y and X scales."
title_hist_fs_logx="Histogram of sgRNA-RSL readcount frequncies in Input library under different filiering cut-offs. Full Y and log(X) scales."
title_hist_lim_logx="Histogram of sgRNA-RSL readcount frequncies in Input library under different filiering cut-offs. Limited Y and log(X) scales to focus on region with most data points."


  hist.fs=ggplot(data=freqs, aes(x=readcount, y=frequency, colour=CO, fill=CO)) + geom_bar(stat="identity") +
  facet_wrap(~CO)+
    scale_fill_viridis(discrete=TRUE,option="turbo")+
    scale_colour_viridis(discrete=TRUE,option="turbo")+
    theme_bw() + theme(plot.margin = unit(c(0.5, -0.5, 0, -0.5), "cm"))+ 
    labs(title=title_hist_fs, y="count",x="reads per RSL-guide")+
    theme(aspect.ratio = 2/(1+sqrt(5)) ) + theme(legend.position="none",legend.margin=unit(c(0, -0.5, 0, -0.5), "cm"))


# in new ggplo2
# still older v in docker, so keeping it as before
# Warning message:
# `legend.margin` must be specified using `margin()`
# For the old behavior use `legend.spacing` 



# log scale X

freqs.n0=freqs[-1,]

hist.fs.logx=ggplot(data=freqs.n0, aes(x=readcount, y=frequency, colour=CO, fill=CO)) + geom_bar(stat="identity") +
  facet_wrap(~CO)+
    scale_fill_viridis(discrete=TRUE,option="turbo")+
    scale_colour_viridis(discrete=TRUE,option="turbo")+
    theme_bw() + theme(plot.margin = unit(c(0.5, -0.5, 0, -0.5), "cm"))+ 
    scale_x_continuous(trans='log10') +
    labs(title=title_hist_fs_logx, y="count",x="reads per RSL-guide")+
    theme(aspect.ratio = 2/(1+sqrt(5)) ) + theme(legend.position="none",legend.margin=unit(c(0, -0.5, 0, -0.5), "cm"))


# set limits
# y_lim is max in readcount range 2 - 50 + 10%
x_lim=50
freqs.2_50=freqs%>%filter(readcount>=2, readcount<=50)
y_lim=round(max(freqs.2_50$frequency)+0.1*(max(freqs.2_50$frequency)))


hist.lim.logx=ggplot(data=freqs.n0, aes(x=readcount, y=frequency, colour=CO, fill=CO)) + geom_bar(stat="identity") +
  facet_wrap(~CO)+
    scale_fill_viridis(discrete=TRUE,option="turbo")+
    scale_colour_viridis(discrete=TRUE,option="turbo")+
    theme_bw() + theme(plot.margin = unit(c(0.5, -0.5, 0, -0.5), "cm"))+ 
    coord_cartesian(xlim=c(1,x_lim),ylim=c(1,y_lim) )+
    scale_x_continuous(trans='log10') +
    labs(title=title_hist_lim_logx, y="count",x="reads per RSL-guide")+
    theme(aspect.ratio = 2/(1+sqrt(5)) ) + theme(legend.position="none",legend.margin=unit(c(0, -0.5, 0, -0.5), "cm"))

```

## Histogram plots

## Histogram plots - full data range


Histogram plots depicing sgRNA-RSL read count frequencies are presented on Figures below. 
Full scales are to present data range, and limited scales (X scale after log10 transformation) to focus on data rich region. 
Limits to Y scale are set to include maximal read count in range 2-50 under all filtering cut offs.



```{r fig.align="center", fig.width=8,fig.height=16, echo = FALSE}
hist.fs
```


<br>
<br>

## Histogram plots - limited data range, X axis log10-transformed


```{r fig.align="center", fig.width=8,fig.height=16, echo = FALSE}
hist.lim.logx
```


<br>
<br>



```{r, eval=refdata.status, child='input_refdata.Rmd'}
```


